ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,563 @@
+    for (AmbryPartition partition : partitionNameToPartitionIdMap.values()) {
+      if (partition.getPartitionState() == PartitionState.READ_WRITE) {
+        writablePartitions.add(partition);
+        if (areAllReplicasForPartitionUp(partition)) {
+          healthyWritablePartitions.add(partition);
+        }
+      }
+    }
+    return healthyWritablePartitions.isEmpty() ? writablePartitions : healthyWritablePartitions;
+  }
+
+  /**
+   * Check whether all replicas of the given {@link AmbryPartition} are up.
+   * @param partition the {@link AmbryPartition} to check.
+   * @return true if all associated replicas are up; false otherwise.
+   */
+  private boolean areAllReplicasForPartitionUp(AmbryPartition partition) {
+    for (AmbryReplica replica : partitionIdToReplicaIds.get(partition)) {
+      if (replica.isDown()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Return the unique {@link AmbryReplica} for a {@link AmbryPartition} on a {@link AmbryDataNode}.
+   * @param hostname the hostname associated with the {@link AmbryDataNode}.
+   * @param port the port associated with the {@link AmbryDataNode}.
+   * @param partitionString the partition id string associated with the {@link AmbryPartition}.
+   * @return the {@link AmbryReplica} associated with the given parameters.
+   */
+  AmbryReplica getReplicaForPartitionOnNode(String hostname, int port, String partitionString) {
+    for (AmbryReplica replica : dataNodeIdToReplicaIds.get(getDataNodeId(hostname, port))) {
+      if (replica.getPartitionId().toString().equals(partitionString)) {
+        return replica;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * An instance of this object is used to register as listner for Helix related changes in each datacenter.
+   */
+  private class ClusterChangeListener extends RoutingTableProvider implements LiveInstanceChangeListener {
+    final Set<String> allInstances = new HashSet<>();
+    private boolean liveInstanceChangeTriggered = false;
+    private boolean externalViewChangeTriggered = false;
+    private boolean configChangeTriggered = false;
+    private CountDownLatch initialized = new CountDownLatch(3);
+
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      logger.info("Live instance change triggered with: " + liveInstances);
+      Set<String> liveInstancesSet = new HashSet<>();
+      for (LiveInstance liveInstance : liveInstances) {
+        liveInstancesSet.add(liveInstance.getInstanceName());
+        instanceNameToDataNodeIdMap.get(liveInstance.getInstanceName()).setState(HardwareState.AVAILABLE);
+      }
+      Set<String> downInstancesSet = Sets.difference(allInstances, liveInstancesSet);
+      for (String instanceName : downInstancesSet) {
+        instanceNameToDataNodeIdMap.get(instanceName).setState(HardwareState.UNAVAILABLE);
+      }
+      if (!liveInstanceChangeTriggered) {
```

Comment:
That is what I was told - but have now added explicit synchronization between so as to guarantee that all the notification methods are serialized - as Helix documentation does not make it clear. If the caller indeed serializes the notifications, then there wouldn't be any contention and the lock would be harmless.